### PR TITLE
For #8310, add nacos-address-plugin module

### PR DESCRIPTION
**Pull request by MRGUOKING**
Please do not create a Pull Request without creating an issue first.

## What is the purpose of the change

For #8310 , add nacos-address-plugin module

## Brief changelog

XX

## Verifying this change

XXXX

Follow this checklist to help us incorporate your contribution quickly and easily:

* [x] Make sure there is a Github issue filed for the change (usually before you start working on it). Trivial changes like typos do not require a Github issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue.
* [x] Format the pull request title like `[ISSUE #123] Fix UnknownException when host config not exist`. Each commit in the pull request should have a meaningful subject line and body.
* [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
* [x] Write necessary unit-test to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add integration-test in [test module](https://github.com/alibaba/nacos/tree/master/test).
* [x] Run `mvn -B clean package apache-rat:check findbugs:findbugs -Dmaven.test.skip=true` to make sure basic checks pass. Run `mvn clean install -DskipITs` to make sure unit-test pass. Run `mvn clean test-compile failsafe:integration-test`  to make sure integration-test pass.

### CODE REVIEW

**Comment by The-Gamer-01** [comment]
Verifying this change需要通过和勾选下面的多选框

File: plugin/address/src/main/java/com/alibaba/nacos/plugin/address/spi/AddressPlugin.java [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+    /**
+     * Get address plugin name.
+     */
+    String getPluginName();
+    
+    /**
+     * When nacos server list change, then call address listener.
+     *@Param: address listener.
+     */
+    void registerListener(Consumer<List<String>> addressListener);
```
Comment (brotherlu-xcq) [inline plugin/address/src/main/java/com/alibaba/nacos/plugin/address/spi/AddressPlugin.java:54]: The return value maked as AddressPlugin is better. The we can do register then start as a chain.

File: plugin/address/src/main/java/com/alibaba/nacos/plugin/address/spi/AbstractAddressPlugin.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    protected Consumer<List<String>> addressListener;
+    
+    @Override
+    public List<String> getServerList() {
+        return serverList;
+    }
+    
+    @Override
+    public void registerListener(Consumer<List<String>> addressListener) {
+        this.addressListener = addressListener;
```
Comment (brotherlu-xcq) [inline plugin/address/src/main/java/com/alibaba/nacos/plugin/address/spi/AbstractAddressPlugin.java:40]: We need to check the if the addressListener is empty.

File: common/src/main/java/com/alibaba/nacos/common/utils/CollectionUtils.java [hunk 3]
```diff
@@ -302,6 +308,61 @@ public static <T> T getOnlyElement(Iterable<T> iterable) {
...
+     * cardinalities.
+     *
+     * <p>That is, iff the cardinality of <i>e</i> in <i>a</i> is equal to the cardinality of <i>e</i> in <i>b</i>, for
+     * each element <i>e</i> in <i>a</i> or <i>b</i>.
+     *
+     * @param a the first collection, must not be null
+     * @param b the second collection, must not be null
+     * @return <code>true</code> iff the collections contain the same elements with the same cardinalities.
+     */
+    public static boolean isEqualCollection(final Collection a, final Collection b) {
```
Comment (brotherlu-xcq) [inline common/src/main/java/com/alibaba/nacos/common/utils/CollectionUtils.java:345]: If you want the collection not null, I think check it before you use it is better.

File: common/src/main/java/com/alibaba/nacos/common/utils/CollectionUtils.java [hunk 4]
```diff
@@ -302,6 +308,61 @@ public static <T> T getOnlyElement(Iterable<T> iterable) {
...
+        if (a.size() != b.size()) {
+            return false;
+        } else {
```
Comment (brotherlu-xcq) [inline common/src/main/java/com/alibaba/nacos/common/utils/CollectionUtils.java:348]: The unnecessary usage of else.

**Review by brotherlu-xcq — CHANGES_REQUESTED** [review]

**Review by KomachiSion — APPROVED** [review]

**Pull request merged**